spec.html

```diff
@@ -24637,17 +25241,50 @@ <h1>
       <emu-alg>
         1. Let _name_ be the result of evaluating |ClassElementName|.
         1. ReturnIfAbrupt(_name_).
+        1. Let _initializers_ be a new empty List.
         1. If |Initializer?| is present, then
-          1. Let _formalParameterList_ be an instance of the production <emu-grammar>FormalParameters : [empty]</emu-grammar>.
-          1. Let _scope_ be the LexicalEnvironment of the running execution context.
-          1. Let _privateScope_ be the running execution context's PrivateEnvironment.
-          1. Let _sourceText_ be the empty sequence of Unicode code points.
-          1. Let _initializer_ be ! OrdinaryFunctionCreate(%Function.prototype%, _sourceText_, _formalParameterList_, |Initializer|, ~non-lexical-this~, _scope_, _privateScope_).
-          1. Perform MakeMethod(_initializer_, _homeObject_).
-          1. Set _initializer_.[[ClassFieldInitializerName]] to _name_.
-        1. Else,
-          1. Let _initializer_ be ~empty~.
-        1. Return the ClassFieldDefinition Record { [[Name]]: _name_, [[Initializer]]: _initializer_ }.
+          1. Let _initializer_ be ! CreateFieldInitializerFunction(_homeObject_, _name_, |Initializer|).
+          1. Append _initializer_ to _initializers_.
+        1. For each element _decorator_ of _decorators_, do
+          1. Let _context_ be CreateDecoratorContextObject(*"field"*, _name_, _metadataList_, ~empty~, _isStatic_).
+          1. Let _initializer_ be ? Call(_decorator_, *undefined*, &laquo; *undefined*, _context_ &raquo;).
+          1. If IsCallable(_initializer_) is *true*, append _initializer_ to _initializers_.
+          1. Else if _initializer_ is not *undefined*, throw a *TypeError* exception.
+        1. Return ClassElementDefinition { [[Key]]: _name_, [[Kind]]: ~field~, [[Initializers]]: _initializers_ }.
+      </emu-alg>
+      <emu-grammar>
+        FieldDefinition : `accessor` ClassElementName Initializer?
+      </emu-grammar>
+      <emu-alg>
+        1. Let _name_ be the result of evaluating |ClassElementName|.
+        1. ReturnIfAbrupt(_name_).
+        1. Let _privateStateDesc_ be the string-concatenation of *"accessor storage"*.
+        1. Let _privateStateName_ be a new Private Name whose [[Description]] value is _privateStateDesc_.
+        1. Let _getter_ be MakeAutoAccessorGetter(_homeObject_, _name_, _privateStateName_).
+        1. Let _setter_ be MakeAutoAccessorSetter(_homeObject_, _name_, _privateStateName_).
+        1. Let _initializers_ be a new empty List.
+        1. If |Initializer?| is present, then
+          1. Let _initializer_ be ! CreateFieldInitializerFunction(_homeObject_, _name_, |Initializer|).
+          1. Append _initializer_ to _initializers_.
+        1. For each element _decorator_ of _decorators_, do
+          1. Let _context_ be CreateDecoratorContextObject(*"accessor"*, _name_, _metadataList_, _extraInitializers_, _isStatic_).
```

Comment:
what's the difference between `extraInitializers` and `initialiers`? Why some functions is appended to the prior one and some are appended to the latter one?

Reply:
Accessors have two initialization steps:

1. The per-instance extraInitializers are run when the class instance is created, _before_ class fields are assigned. This is the same time that method initializers in general are run.
2. The private field backing the accessor is initialized. This runs during class field initialization.

Accessors are essentially sugar for declaring a getter/setter + field, so these two steps can be thought of as the combination of the initialization for both of those types of values, e.g.:

```js
class Foo {
  @dec
  #bar = 123;

  @dec
  get bar() {
    return this.#bar;
  }
}
```

In this example, `@dec` would be able to add an initializer for the `#bar` private field which would run when the field is defined, and separately add an initializer for the `get bar()` getter which would run at the beginning of class instantiation. Decorating an `accessor` lets you do both.